js/private/js_run_devserver.js

```diff
@@ -0,0 +1,165 @@
+const path = require('path')
+const fs = require('fs')
+const os = require('os')
+const child_process = require('child_process')
+
+const RUNFILES_ROOT = path.join(
+    process.env.RUNFILES,
+    process.env.JS_BINARY__WORKSPACE
+)
+
+const synced = new Map()
+
+async function mkdirp(p) {
+    if (p && !fs.existsSync(p)) {
+        await mkdirp(path.dirname(p))
+        await fs.promises.mkdir(p)
+    }
+}
+
+function isExecutable(stats) {
+    return (
+        !!(stats.mode & fs.constants.S_IXUSR) ||
+        !!(stats.mode & fs.constants.S_IXGRP) ||
+        !!(stats.mode & fs.constants.S_IXOTH)
+    )
+}
+
+async function syncRecursive(src, dst) {
+    try {
+        const lstat = await fs.promises.lstat(src)
+        const last = synced.get(src)
+        if (!lstat.isDirectory() && last && lstat.mtimeMs == last) {
+            // this file is already up-to-date
+            return
+        }
+        synced.set(src, lstat.mtimeMs)
```

Comment:
This never gets cleaned so it will grow forever if files are added/removed. Do we even need it? It seems like it just avoids an `lstat(dst)` which can replace the `existsSync` below anyway...?

```
const [srcLstat, destLstat] = await Promise.all([
   fs.promises.lstat(src),
   fs.promises.lstat(dst).catch(() => null), //<-- .catch() required for file not existing case?
])
```
if (!srcLstat.isDirectory() && destLstat && srcLstat.mtimeMs === destLstat.mtimeMs) return;

const exists = !!destLstat;

Reply:
It is bounded by the set of files in config.data_files. We do need the map of file => mtimeMS to know which files have changed since the last sync. On every sync we go through the full set of data_files and copy over any files that have changed or are new since the last sync.

Reply:
But `config.data_files` can continue to change and `synced` never removes anything. If I just rename a file the old one will never be removed, although that might be small enough who cares? 🤷

Reply:
> We do need the map of file => mtimeMS to know which files have changed since the last sync.

Can't we do `lstat` to find which files have changed since the last sync? It gives you all the same information but avoids a) the map, b) the `existsSync` down a line. However that `lstat` would then be *before* the `if (changed)` where the `existsSync` currently is *after*.

Reply:
Yes, technically if you keep adding files to your BUILD graph then it grows unbounded... tho it is only during the course of an ibazel session and your output tree would grow unbounded as well in that case. It wouldn't have hard to check what files are no longer present and unlink them from the tmpdir and remove then from the set tho. Might add that before landing for completeness. I wonder if the runfiles tree also grows unbounded? Assuming it does and it is not wiped clean between builds.

Reply:
Do you mean we can check the timestamp from lstat against the time of the last sync for all files? There are subtle race conditions with that approach since a sync goes over a timespan from the first copy to the last so its trick to get a timestamp for a sync that works for all files.

Reply:
> I wonder if the runfiles tree also grows unbounded?

I'd hope it cleans deleted files, otherwise it's not properly sandboxed?

> Might add that before landing for completeness.

It might be true that it grows forever but it's also possible it's not worth doing anything. With what I suggested we wouldn't need it, with the current approach it might not be worth it just to save some tiny map entries in a process that probably won't live long enough for it to matter?

> Do you mean we can check the timestamp from lstat against the time of the last sync for all files?

It would be checking the lastmod time of the src + dst at the time of copy. The lastmod of the dst should be the same as what's in the map... unless a different process touched it.

Reply:
Ohhh. I see the source of confusion. The `lstat.mtimeMs` of the dst is the time the copy is made and will never be the same as the `lstat.mtimeMs` of the src. `copy` doesn't preserve the `lstat.mtimeMs`

Reply:
Hence the map to keep track of the `lstat.mtimeMs` of the src when we do the copy

Reply:
Ohhh I see. We're storing+comparing the time of the src, not comparing the src vs destination 👍

Reply:
So the map is required. I don't think it's wroth trying to clean it atm, at least in this PR

Reply:
Yeah. Probably not worth the extra complexity. Will leave cleanup out for now.